Add BooleanArray::binary/unary helpers backed by buffer bitwise APIs

joe-ucp opened the pull request:
## What issue does this PR close?

- **Part of**: #8806 (Bitwise API consolidation).

This PR extracts the BooleanArray bitwise changes from a larger exploratory PR to enable focused and independent review.

---

## Rationale

This change consolidates BooleanArray bitwise logic by leveraging new buffer-level bitwise primitives. By doing so:

- **High-level kernels become simpler** and easier to maintain.
- **Bitmap behavior** (including offsets, tail bits, and nulls) can be controlled and tested in one place.

---

## Summary of Changes

- **New Helper Methods**  
  - Add `BooleanArray::binary` and `BooleanArray::unary` methods to operate on the value bitmap, built on top of `Buffer::bitwise_binary` and `Buffer::bitwise_unary`.
- **Kernel Updates**  
  - Update the boolean kernels in `arrow-arith/src/boolean.rs` to use these new helpers instead of manual bitmap logic.
- **New Tests**  
  - Add property-based tests in `arrow-arith` to validate the new helpers against existing logic across a range of offsets and lengths.
  - Add `rand` as a dev-dependency to `arrow-arith` for randomized testing.
- **No Change To**  
  - No updates to `nullif` or other higher-level APIs in this PR.

---

## Testing

-  `cargo test -p arrow-array --lib`
-  `cargo test -p arrow-arith --lib`
-  `cargo test -p arrow-select --lib`

All tests pass.

---

## User-Facing Changes

**None.**  
This is an internal refactor: new helpers and tests only. No changes to existing public APIs or observable behavior.

joe-ucp committed:
Add BooleanArray bitwise unary/binary APIs and tests

Add BooleanArray::binary and BooleanArray::unary in arrow-array, implemented in terms of Buffer::bitwise_binary / Buffer::bitwise_unary

Update boolean kernels in arrow-arith to use these helpers instead of open-coded bitmap logic

Add randomized equivalence tests in arrow-arith to validate the new APIs against existing behavior

Add rand as a dev-dependency in arrow-arith for the new tests

All behavior should be unchanged; this is an internal refactor toward bitwise API consolidation

Tests run locally:

cargo test -p arrow-array --lib

cargo test -p arrow-arith --lib

cargo test -p arrow-select --lib

alamb reviewed (commented):
Thanks @joe-ucp  -- this looks like a good start.

joe-ucp committed:
Use FnMut for BooleanArray unary/binary and restore buffer doc example

Align BooleanArray helper bounds with Buffer bitwise APIs and restore the removed doc example in immutable.rs.
